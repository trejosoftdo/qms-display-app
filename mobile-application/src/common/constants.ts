export const AUTHORIZATION_PENDING_CODE = 'authorization_pending';

export const VALIDATION_ERROR_STATUS_CODE = 400;

export const DEVICE_NOT_CONNECTED_ERROR_MESSAGE = 'Device is not connected';

export const CONNECTION_EXPIRED_ERROR_MESSAGE = 'Connection expired!';

export const APP_PROTO = 'app://';

export const DEVICE_ID_KEY = 'SECURE_DEVICE_ID';

export const CONNECTION_DETAILS_KEY = 'CONNECTION_DETAILS';

export const WEB_PLATFORM = 'web';

export const BEARER_PORTION = 'Bearer';

export const EMPTY_VALUE = '';

export const DEFAULT_REGULAR_FONT_NAME = 'Montserrat-Regular';

export const DEFAULT_LIGHT_FONT_NAME = 'Montserrat-Light';

export const DEFAULT_MEDIUM_FONT_NAME = 'Montserrat-Medium';

export const NORMAL_FONT_WEIGHT = 'normal';

export const ONE_SECOND_MILISECONDS = 1000;

export const INTERVAL_TIME = 10 * ONE_SECOND_MILISECONDS;

export const SERVICES_PATH = '/services';

export const TICKET_DETAILS_PATH = '/ticket-details';

export const HOME_PATH = '/';

export const CATEGORIES_PATH = '/categories';

export const CONFIGURATION_PATH = '/configuration';

export const NO_OP = () => {};


export const LANGUAGES = {
  SPANISH: 'es',
  ENGLISH: 'en',
};

export const CONFIG_OPTIONS = {
  INFORMATION: 'information',
  CONNECTION: 'connection',
  LANGUAGE: 'language',
};
