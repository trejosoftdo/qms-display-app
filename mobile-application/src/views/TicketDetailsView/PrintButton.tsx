import * as React from 'react';
import { useTranslation } from 'react-i18next';
import ActionButton from '../../common/components/ActionButton';
import { TRANSLATION_PRINT_TICKET_KEY } from '../../common/translations/translation-keys';

/**
 * PrintButtonProps defines the props for the Printing Form Component.
 *
 * @property onPress - on press handler
 */
interface PrintButtonProps {
  onPress: () => void;
}

/**
 * A component that represents a button for printing ticket details
 *
 * @param {PrintButtonProps} props - The props for the Print Button component.
 */
const PrintButton: React.FC<PrintButtonProps> = (props: PrintButtonProps) => {
  const { t } = useTranslation();
  return (
    <ActionButton
      icon="printer"
      onPress={props.onPress}
      message={t(TRANSLATION_PRINT_TICKET_KEY)}
    />
  );
};

export default PrintButton;
