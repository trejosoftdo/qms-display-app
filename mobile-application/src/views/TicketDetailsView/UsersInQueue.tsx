import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { StyleSheet } from 'react-native';
import { Text } from 'react-native-paper';
import { TRANSLATION_USERS_IN_QUEUE_KEY } from '../../common/translations/translation-keys';


/**
 * UsersInQueueProps defines the props for the Users In Queue Component.
 *
 * @property total - total of users in queue
 */
interface UsersInQueueProps {
  total: number;
};

/**
 * A component to display the number of users who are in queue
 *
 * @param {UsersInQueueProps} props - The props for the Users In Queue component.
 */
const UsersInQueue: React.FC<UsersInQueueProps> = (props: UsersInQueueProps) => {
  const { t } = useTranslation();
  return (
    <Text style={styles.container} variant="titleMedium">
      {t(TRANSLATION_USERS_IN_QUEUE_KEY, { total: props.total })}
    </Text>
  );
};

const styles = StyleSheet.create({
  container: {
    paddingBottom: 32,
  },
});

export default UsersInQueue;
