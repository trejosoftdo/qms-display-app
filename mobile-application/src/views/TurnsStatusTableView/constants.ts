export const BEING_ATTENDED_STATUS_CODE = "BEING_ATTENDED";

export const TO_BE_ATTENDED_STATUS_CODE = "TO_BE_ATTENDED";