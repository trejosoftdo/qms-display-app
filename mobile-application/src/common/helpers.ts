import { Platform } from 'react-native';
import { router } from 'expo-router';
import { APP_PROTO, EMPTY_VALUE, ONE_SECOND_MILISECONDS, WEB_PLATFORM } from './constants';

/** 
 * Checks if the platform is web
 * @returns boolean
 */
export const isWeb = (): boolean => Platform.OS === WEB_PLATFORM;

/** 
 * Gets the current time
 * @returns number
 */
export const getCurrentTime = (): number => new Date().getTime();

/**
 * Calculates the expire time
 * 
 * @param  {number} expiresIn
 * @returns number
 */
export const calculateExpireTime = (expiresIn: number): number => getCurrentTime() + expiresIn * ONE_SECOND_MILISECONDS;


/** 
 * Extracs icon name
 * @returns string
 */
export const extractIconName = (iconUrl: string): string => iconUrl.replace(APP_PROTO, EMPTY_VALUE)

/**
 * Navigates to a given path
 * 
 * @param  {string} pathname
 * @param  {any} params
 * @returns void
 */
export const goToPath = (pathname: string, params: Record<string, any> = undefined): void => {
  router.push({
    pathname,
    params,
  });
};

/**
 * Waits for certain time
 * @param  {number} timeout 
 * @returns Promise<any>
 */
export const delay = (timeout: number) => new Promise((resolve) => {
  setTimeout(resolve, timeout);
});

/**
 * Gets the response blob url
 * @param  {Response} response 
 * @returns Promise<string>
 */
export const getResponseBlobUrl = async (response: Response): Promise<string> => {
  const stream = await response.blob();
  return URL.createObjectURL(stream);
};
