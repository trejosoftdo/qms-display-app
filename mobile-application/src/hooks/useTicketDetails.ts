import { useProgress } from "../common/hooks";
import { Progress, TicketDetailsData } from "../common/models";
import { loadTicketDetails } from "../common/services/ticket";

/**
 * Hook used to get the details of a ticket for a given service.
 * @param  {string} service
 * 
 * @returns Progress<TicketDetailsData>
 */
const useTicketDetails = (service: string, customerName: string): Progress<TicketDetailsData> => useProgress<TicketDetailsData>(loadTicketDetails(service, customerName));

export default useTicketDetails;
