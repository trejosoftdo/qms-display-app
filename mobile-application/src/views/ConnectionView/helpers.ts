import { ConnectionData, ConnectionDetails } from "../../common/models";
import { TRANSLATION_CONNECTED_MESSAGE_KEY, TRANSLATION_INITIAL_CONNECT_MESSAGE_KEY, TRANSLATION_SUCCESS_CONNECTION_MESSAGE_KEY, TRANSLATION_TO_CONNECT_MESSAGE_KEY } from "../../common/translations/translation-keys";

/**
 * Gets the correct message key from the data state
 * 
 * @param  {ConnectionData} data
 * @returns string
 */
export const getMessageKey = (data: ConnectionData, connectionDetails?: ConnectionDetails): string => {
  if (connectionDetails?.accessToken) {
    return TRANSLATION_CONNECTED_MESSAGE_KEY;
  }

  if (data?.tokens?.accessToken) {
    return TRANSLATION_SUCCESS_CONNECTION_MESSAGE_KEY;
  }

  return data?.userCode ? TRANSLATION_TO_CONNECT_MESSAGE_KEY : TRANSLATION_INITIAL_CONNECT_MESSAGE_KEY;
};
