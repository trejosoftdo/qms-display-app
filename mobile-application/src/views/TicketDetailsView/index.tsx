import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocalSearchParams } from 'expo-router';
import { HOME_PATH } from '../../common/constants';
import { TRANSLATION_WAIT_MESSAGE_KEY, TRANSLATION_WELCOME_TURN_KEY } from '../../common/translations/translation-keys';
import { goToPath } from '../../common/helpers';
import { AppView, ConditionalContainer, Value } from '../../common/components';
import useTicketDetails from '../../hooks/useTicketDetails';
import UsersInQueue from './UsersInQueue';
import PrintButton from './PrintButton';

/**
 * TicketDetailsViewProps defines the props for the Ticket Details View Component.
 */
interface TicketDetailsViewProps {}

/**
 * A component for the Ticket Details view
 *
 * @param {TicketDetailsViewProps} props - The props for the Ticket Details View component.
 */
const TicketDetailsView: React.FC<TicketDetailsViewProps> = (props: TicketDetailsViewProps) => {
  const params = useLocalSearchParams();
  const {
    loading,
    data,
    error,
  } = useTicketDetails(params.serviceId.toString(), params.customerName.toString());
  const { t } = useTranslation();
  const messageKey = loading ? TRANSLATION_WAIT_MESSAGE_KEY : TRANSLATION_WELCOME_TURN_KEY;
  return (
    <AppView
      headerMessage={t(messageKey, { service: params.serviceLabel.toString(), customer: params.customerName.toString() })}
      loading={loading}
      error={error}
    >
      <ConditionalContainer display={!loading && !!data}>
        <>
          <Value
            value={data?.details.value}
            icon="cards-outline"
          />
          <UsersInQueue total={data?.usersInQueue} />
          <PrintButton
            onPress={() => {
              goToPath(HOME_PATH);
            }}
          />
        </>
      </ConditionalContainer>
    </AppView>
  );
};

export default TicketDetailsView;
