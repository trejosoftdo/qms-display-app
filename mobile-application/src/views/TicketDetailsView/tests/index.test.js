import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { useLocalSearchParams } from 'expo-router';
import { goToPath } from '../../../common/helpers';
import { HOME_PATH } from '../../../common/constants';
import useTicketDetails from '../../../hooks/useTicketDetails';
import TicketDetailsView from '../';


jest.mock('expo-router');
jest.mock('../../../common/helpers');
jest.mock('../../../hooks/useTicketDetails');


describe('TicketDetailsView component', () => {
  const mockParams = {
    customerName: 'customer-name',
    serviceLabel: 'mock-service-label',
    serviceId: '1234',
  };
  const mockData = {
    details: {
      value: 'mock-value',
    },
    usersInQueue: 14,
  };

  beforeEach(() => {
    useLocalSearchParams.mockReturnValue(mockParams);
  });

  afterEach(() => {
    jest.clearAllMocks();
  })

  it('renders the component as expected when loading', () => {
    useTicketDetails.mockReturnValue({ loading: true });
    expect(renderShallow(<TicketDetailsView />)).toMatchSnapshot();
    expect(goToPath).toHaveBeenCalledTimes(0);
    expect(useLocalSearchParams).toHaveBeenCalledTimes(1);
    expect(useLocalSearchParams).toHaveBeenCalledWith();
    expect(useTicketDetails).toHaveBeenCalledTimes(1);
    expect(useTicketDetails).toHaveBeenCalledWith(
      mockParams.serviceId,
      mockParams.customerName,
    );
  });

  it('renders the component as expected when data has been loaded', () => {
    useTicketDetails.mockReturnValue({
      loading: false,
      data: mockData,
    });
    expect(goToPath).toHaveBeenCalledTimes(0);
    expect(renderShallow(<TicketDetailsView />)).toMatchSnapshot();
    expect(useLocalSearchParams).toHaveBeenCalledTimes(1);
    expect(useLocalSearchParams).toHaveBeenCalledWith();
    expect(useTicketDetails).toHaveBeenCalledTimes(1);
    expect(useTicketDetails).toHaveBeenCalledWith(
      mockParams.serviceId,
      mockParams.customerName,
    );
  });

  it('renders the component as expected when it has errors', () => {
    useTicketDetails.mockReturnValue({
      loading: false,
      error: new Error('Unexpected error'),
    });
    expect(renderShallow(<TicketDetailsView />)).toMatchSnapshot();
    expect(goToPath).toHaveBeenCalledTimes(0);
    expect(useLocalSearchParams).toHaveBeenCalledTimes(1);
    expect(useLocalSearchParams).toHaveBeenCalledWith();
    expect(useTicketDetails).toHaveBeenCalledTimes(1);
    expect(useTicketDetails).toHaveBeenCalledWith(
      mockParams.serviceId,
      mockParams.customerName,
    );
  });

  it('goes to services path when an item is selected', () => {
    useTicketDetails.mockReturnValue({
      loading: false,
      data: mockData,
    });
    const { getByRole } = render(<TicketDetailsView />);
    const element = getByRole('button');

    fireEvent.press(element);

    expect(goToPath).toHaveBeenCalledTimes(1);
    expect(goToPath).toHaveBeenCalledWith(HOME_PATH);
    expect(useLocalSearchParams).toHaveBeenCalledTimes(1);
    expect(useLocalSearchParams).toHaveBeenCalledWith();
    expect(useTicketDetails).toHaveBeenCalledTimes(1);
    expect(useTicketDetails).toHaveBeenCalledWith(
      mockParams.serviceId,
      mockParams.customerName,
    );
  });
});
