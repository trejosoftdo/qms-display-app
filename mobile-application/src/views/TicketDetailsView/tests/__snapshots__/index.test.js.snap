// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TicketDetailsView component renders the component as expected when data has been loaded 1`] = `
<AppView
  headerMessage="Translated[translation:welcomeTurn]({"service":"mock-service-label","customer":"customer-name"})"
  loading={false}
>
  <ConditionalContainer
    display={true}
  >
    <React.Fragment>
      <Value
        icon="cards-outline"
        value="mock-value"
      />
      <UsersInQueue
        total={14}
      />
      <PrintButton
        onPress={[Function]}
      />
    </React.Fragment>
  </ConditionalContainer>
</AppView>
`;

exports[`TicketDetailsView component renders the component as expected when it has errors 1`] = `
<AppView
  error={[Error: Unexpected error]}
  headerMessage="Translated[translation:welcomeTurn]({"service":"mock-service-label","customer":"customer-name"})"
  loading={false}
>
  <ConditionalContainer
    display={false}
  >
    <React.Fragment>
      <Value
        icon="cards-outline"
      />
      <UsersInQueue />
      <PrintButton
        onPress={[Function]}
      />
    </React.Fragment>
  </ConditionalContainer>
</AppView>
`;

exports[`TicketDetailsView component renders the component as expected when loading 1`] = `
<AppView
  headerMessage="Translated[translation:waitMessage]({"service":"mock-service-label","customer":"customer-name"})"
  loading={true}
>
  <ConditionalContainer
    display={false}
  >
    <React.Fragment>
      <Value
        icon="cards-outline"
      />
      <UsersInQueue />
      <PrintButton
        onPress={[Function]}
      />
    </React.Fragment>
  </ConditionalContainer>
</AppView>
`;
